fillers/eips/eip4844/datahash_opcode.py

```diff
@@ -538,118 +545,238 @@ def test_datahash_gas(_: str):
+    post = {}
+
+    datahash_verbatim = "verbatim_{}i_{}o".format(
+        Op.DATAHASH.popped_stack_items, Op.DATAHASH.pushed_stack_items
     )
 
-    tx2 = Transaction(
-        ty=2,
-        nonce=1,
-        data=to_hash_bytes(0),
-        to=datahash_max_gas_measure_code_address,
-        gas_limit=3000000,
-        max_fee_per_gas=10,
-        max_priority_fee_per_gas=10,
-        access_list=[],
+    # Start from -INVALID_DEPTH to (${num_blobs} + INVALID_DEPTH)
+    datahash_invalid_index_code = Template(
+        f"""
+        {{
+            let pos := 0
+            let relative_pos := sub(0, {INVALID_DEPTH})
+            let end := add(${{num_blobs}}, {INVALID_DEPTH})
+            for {{}} lt(pos, end) {{ pos := add(pos, 1) }}
+            {{
+                let adjusted_pos := add(pos, relative_pos)
+                let datahash := {datahash_verbatim}(
+                    hex"{Op.DATAHASH.hex()}", adjusted_pos
+                )
+                sstore(adjusted_pos, datahash)
+            }}
+        }}
+        """
     )
 
-    post = {
-        datahash_zero_gas_measure_code_address: Account(
+    for i in range(NUM_BLOCKS):
+        address = to_address(0x100 + i * 0x100)
+        rnd_blob_size = random.randint(0, MAX_BLOB_PER_BLOCK)
+        pre[address] = Account(
+            code=Yul(
+                datahash_invalid_index_code.substitute(num_blobs=rnd_blob_size)
+            )
+        )
+        blob_ver_hashes = [
+            to_hash_bytes(random.randint(MAX_BLOB_PER_BLOCK, 2**256 - 1))
+            for _ in range(rnd_blob_size)
+        ]
+        blocks.append(
+            Block(
+                txs=[  # Create tx with rnd num valid blobs per block
+                    Transaction(
+                        ty=5,
+                        nonce=i,
+                        data=to_hash_bytes(0),
+                        to=address,
+                        gas_limit=3000000,
+                        max_fee_per_gas=10,
+                        max_priority_fee_per_gas=10,
+                        max_fee_per_data_gas=10,
+                        access_list=[],
+                        blob_versioned_hashes=blob_ver_hashes,
+                    )
+                ]
+            )
+        )
+        post[address] = Account(
             storage={
-                0: DATAHASH_GAS_COST,
+                index: 0
+                if index < 0 or index >= rnd_blob_size
+                else blob_ver_hashes[index]
+                for index in range(
+                    -INVALID_DEPTH, rnd_blob_size + INVALID_DEPTH
+                )
             }
-        ),
-        datahash_max_gas_measure_code_address: Account(
-            storage={
-                0: DATAHASH_GAS_COST,
-            }
-        ),
-    }
+        )
 
-    yield StateTest(
-        env=env,
+    yield BlockchainTest(
         pre=pre,
         post=post,
-        txs=[tx1, tx2],
-        tag="cost_tx_type_2",
+        blocks=blocks,
     )
```

Comment:
We should also test opcodes that access tx information, since type 5 tx is now SSZ, there could be issues there. E.g.
- ORIGIN
- GASPRICE
- CHAINID
- CALLER (On the top level call stack)
- CALLVALUE (On the top level call stack)